seed/seed20.go

```diff
@@ -96,6 +96,103 @@ type seed20 struct {
 	essentialSnapsNum int
 }
 
+// OptionalContainers implements the Copier interface.
+func (s *seed20) OptionalContainers() (OptionalContainers, error) {
+	model := s.Model()
+
+	optionalSnaps := make(map[string]bool)
+	optionalComponents := make(map[string]map[string]bool)
+	for _, snap := range model.AllSnaps() {
+		optionalComps := make(map[string]bool)
+		for compName, comp := range snap.Components {
+			if comp.Presence == "optional" {
+				optionalComps[compName] = true
+			}
+		}
+
+		optionalComponents[snap.Name] = optionalComps
+
+		if snap.Presence == "optional" {
+			optionalSnaps[snap.Name] = true
+		}
+	}
+
+	availableSnaps, availableComponents, err := s.availableContainers()
+	if err != nil {
+		return OptionalContainers{}, err
+	}
+
+	available := OptionalContainers{
+		Components: make(map[string][]string),
+	}
+	for sn := range availableSnaps {
+		if optionalSnaps[sn] {
+			available.Snaps = append(available.Snaps, sn)
+		}
+	}
+
+	for sn, comps := range availableComponents {
+		optionalCompsForSnap := optionalComponents[sn]
+		if optionalCompsForSnap == nil {
+			continue
+		}
+
+		for c := range comps {
+			if optionalCompsForSnap[c] {
+				available.Components[sn] = append(available.Components[sn], c)
+			}
+		}
+	}
+
+	if len(available.Components) == 0 {
+		available.Components = nil
+	}
+
+	return available, nil
+}
+
+func (s *seed20) availableContainers() (map[string]bool, map[string]map[string]bool, error) {
+	assertedSnapSet, assertedCompSets := s.availableAssertedContainers()
+
+	optsPath := filepath.Join(s.systemDir, "options.yaml")
+	if osutil.FileExists(optsPath) {
```

Comment:
we probably shouldn't even check for this if the grade of the model is not right? is that the logic that LoadMeta has?